### PR TITLE
fix(datepicker): solve refocusing problems

### LINKED ISSUE #8960: angular material (datepicker) focus is on body after any keyboard based operation
**Actual Behavior**:

Its seems that ever keyboard based operation with the date-picker don't return the focus back to the date picker input or button.
As a result the tab indexing is back to the first element. 

you can take example with the official demo:
looking at the sources and debugging - its seems that focus operation is triggered on the relevant element but even so - the focus changed to to the document body.

**CodePen**
https://material.angularjs.org/latest/demo/datepicker

**Angular Versions**:  - latest 

**Additional Information**:
 -  `Browser Type: *` chrome

### PR DESCRIPTION
**Pull request by crisbeto**
* Fixes the datepicker not being able to refocus on the calendar icon, because the SVG element inside didn't propagate the focus event.
* Fixes the datepicker sometimes not refocusing on the element that triggered it.

Fixes #8960.

### CODE REVIEW

File: src/components/datepicker/js/datepickerDirective.js [hunk 1]
```diff
@@ -677,28 +677,29 @@
     if (this.isCalendarOpen) {
       var self = this;
 
+      self.detachCalendarPane();
+      self.ngModelCtrl.$setTouched();
+
+      self.documentElement.off('click touchstart', self.bodyClickHandler);
```
Comment (ThomasBurleson) [inline src/components/datepicker/js/datepickerDirective.js]: Why was this moved outside the detach/reset ?

Comment (crisbeto) [inline src/components/datepicker/js/datepickerDirective.js]: Mainly because they don't _need_ to be in it. The only one that should fire on the next tick is the `calendarOpen` since it prevents the calendar reopening immediately if the user turned on `openOnFocus`.
